src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -567,6 +567,24 @@ public T[] ToArray()
         }
 
         /// <summary>
+        /// Removes all values from the <see cref="ConcurrentBag{T}"/>.
+        /// </summary>
+        public void Clear()
+        {
+            // Short path if the bag is empty
+            if (_headList == null)
+                return;
+
+            // Acquire the global lock to update the pointers
+            lock (GlobalListsLock)
+            {
+                _locals = new ThreadLocal<ThreadLocalList>();
+                _headList = null;
+                _tailList = null;
```

Comment:
No. By doing it the way you're doing it, Clear can end up doing only a partial clear, where it can still be clearing from part of the data structure while new elements are being added elsewhere that won't be cleared; you can also end up in situations where you'll end up clearing elements added after the Clear operation completes. This is why you need to take all the locks, not just the global one (which freezing also takes). Plus, doing it that way means you don't have to throw away the thread local and all the non-trivial data structures it references.